### PR TITLE
Add concurrent op check

### PR DESCRIPTION
**Pull request by WayneFerrao**
## Description
[AB#1092](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/1092)
This PR adds assertions to DeltaManager and ContainerRuntime to ensure that ops are not being sent while currently being processsed. Currently this is happening unchecked, leading to non-deterministic outcomes.

### CODE REVIEW

File: packages/loader/container-loader/src/deltaManager.ts [hunk 1]
```diff
@@ -186,6 +190,9 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
     public get clientDetails() { return this.connectionManager.clientDetails; }
 
     public submit(type: MessageType, contents: any, batch = false, metadata?: any) {
+        if (this.opsCurrentlyProcessing > 0) {
+            this.close(new UsageError("Currently processing ops: Container closed"));
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:194]: Let's add better message here. I'd say "Making changes to data model is disallowed while processing ops."

File: packages/loader/container-loader/src/deltaManager.ts [hunk 2]
```diff
@@ -762,6 +768,7 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
 
     private processInboundMessage(message: ISequencedDocumentMessage): void {
         const startTime = Date.now();
+        this.opsCurrentlyProcessing++;
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:771]: I'd rather make it a boolean (as there is no reentrancy here). You could also add asserts that it is false (zero in current form) before we change it here, and true below we reset it back

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 3]
```diff
@@ -1890,6 +1890,10 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
 
     public process(messageArg: ISequencedDocumentMessage, local: boolean) {
         this.verifyNotClosed();
+        if (this.opsCurrentlyProcessing !== 0) {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1893]: I do not remember why we decided to go with checking it in two places, but worth adding some comment on why we are doing

File: packages/loader/container-loader/src/test/deltaManager.spec.ts [hunk 4]
```diff
@@ -332,6 +332,12 @@ describe("Loader", () => {
                     deltaManager.connectionManager.forceReadonly(true);
                 });
             });
+
```
Comment (vladsud) [inline packages/loader/container-loader/src/test/deltaManager.spec.ts:335]: You could test it at that level (at least the check in DeltaManager, so yes, adding some code to do so would be nice).

File: packages/loader/container-loader/src/deltaManager.ts [hunk 5]
```diff
@@ -47,6 +47,7 @@ import {
     DataCorruptionError,
     extractSafePropertiesFromMessage,
     DataProcessingError,
+    UsageError,
 } from "@fluidframework/container-utils";
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:51]: You would probably want to start with only containerRuntime.ts changes (i.e. remove deltaMnager.ts changes) as it's much harder to make them optional.
The way containerRuntime.ts is implemented - it gives a way for people to disable it, at least temporarily and fix their issues outside the process of picking latest version of FF

**Comment by vladsud** [comment]
Please add breaking.md note about these changes, how users can temporarily disable it, and the fact that they will lose such capability in the future, thus need to fix any bugs it finds.

File: packages/runtime/container-runtime/src/test/containerRuntime.spec.ts [hunk 6]
```diff
@@ -518,6 +518,26 @@ describe("Runtime", () => {
...
+                    const message: Partial<ISequencedDocumentMessage> = {
+                        clientId,
+                        type: MessageType.Operation,
+                    };
+
+                    await deltaManager.inbound.pause();
+                    pushOp(message);
+                    pushOp(message);
+                    pushOp(message);
+                    await processOps();
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:532]: I'm not sure what this test actually testing. It does not result in ops being generated while processing an op.
I.e. it will either pass of fail in the same way, with or without your other changes.
Plus, if I got it right, this code uses MockDeltaManager, not real (production) code you are modifying, and there is no ContainerRuntime created in this test (as far as I can see).

I think the easiest way to validate new behavior is to use some existing DDS (in end-to-end tests) and change it from within event handler.

File: packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts [hunk 7]
```diff
@@ -375,6 +389,25 @@ describeNoCompat("SharedMap orderSequentially", (getTestObjectProvider) => {
         });
     });
 
+    itExpects.skip("Should close container when sending an op while processing another op",
```
Comment (andre4i) [inline packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts:392]: `skip`?

File: packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts [hunk 8]
```diff
@@ -375,6 +389,25 @@ describeNoCompat("SharedMap orderSequentially", (getTestObjectProvider) => {
...
+        [{
+            eventName: "fluid:telemetry:Container:ContainerClose",
+            error: "Making changes to data model is disallowed while processing ops.",
+        }], async () => {
+            sharedMap.on("valueChanged", (changed, local) => {
+                if (!local) {
+                    assert.equal(changed.key, "key2", "Incorrect value for key1 in container 1");
```
Comment (andre4i) [inline packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts:399]: what contract is this assertion enforcing related to the current test?

File: packages/loader/container-loader/src/deltaManager.ts [hunk 9]
```diff
@@ -89,6 +90,10 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
     private pending: ISequencedDocumentMessage[] = [];
     private fetchReason: string | undefined;
 
+    // A boolean used to assert that ops are not being sent while processing another op.
+    private opsCurrentlyProcessing: boolean = false;
+    // Delta manager option enabling the above assertion
+    private readonly preventConcurrentOpSend: boolean = false;
```
Comment (andre4i) [inline packages/loader/container-loader/src/deltaManager.ts:96]: this probably needs to be read from a feature gate otherwise it's going to stay disabled until we write the code to enable it, right?

Follow how `maxOpSizeInBytesKey` is used inside `containerruntime.ts` as an example about feature gates.

File: packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts [hunk 10]
```diff
@@ -375,6 +389,25 @@ describeNoCompat("SharedMap orderSequentially", (getTestObjectProvider) => {
         });
     });
 
+    itExpects.skip("Should close container when sending an op while processing another op",
+        [{
+            eventName: "fluid:telemetry:Container:ContainerClose",
+            error: "Making changes to data model is disallowed while processing ops.",
```
Comment (andre4i) [inline packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts:395]: how would this work if `preventConcurrentOpSend` is set to false?

File: BREAKING.md [hunk 11]
```diff
@@ -898,7 +902,7 @@ The `SharedNumberSequence` and `SharedObjectSequence` have been deprecated and a
 Additionally, `useSyncedArray()` from `@fluid-experimental/react` has been removed, as it depended on the `SharedObjectArray`.
 
 ### `IContainer` interface updated to complete 0.53 changes
-The breaking changes introduced in [`IContainer` interface updated to expose actively used `Container` public APIs](#IContainer-interface-updated-to-expose-actively-used-Container-public-APIs) have now been completed in 0.54. The following additions to the `IContainer` interface are no longer optional but rather mandatory:
+The breaking changes introduced in [`IContainer` interface updated to expose actively used `Container` public APIs](#IContainer-interface-updated-to-expose-actively-used-Container-public-APIs) have now been completed in 0.54. The following additions ,to the `IContainer` interface are no longer optional but rather mandatory:
```
Comment (andre4i) [inline BREAKING.md:905]: is this change intentional?

File: BREAKING.md [hunk 12]
```diff
@@ -2556,11 +2560,6 @@ ErrorType enum has been broken into 3 distinct enums / layers:
 
 Due to a change in the sequence's snapshot format clients running a version less than 0.19 will not be able to load snapshots generated in 0.21. This will affect all sequence types includes shared string, and sparse matrix. If you need to support pre-0.19 clients please contact us for mitigations.
 
-### ITelemetryBaseLogger.supportsTags deleted
```
Comment (andre4i) [inline BREAKING.md:2559]: I think this file was out of sync

File: BREAKING.md [hunk 13]
```diff
@@ -23,25 +23,29 @@ It's important to communicate breaking changes to our stakeholders. To write a g
 - [Remove `type` field from `ShareLinkInfoType`](#Remove-type-field-from-ShareLinkInfoType)
 - [Remove `ShareLinkTypes` interface](#Remove-ShareLinkTypes-interface)
 - [Remove `enableShareLinkWithCreate` from `HostStoragePolicy`](#Remove-enableShareLinkWithCreate-from-HostStoragePolicy)
+- [Add assertion that closes container in DeltaManager if sending an op while concurrently processing another op](#add-assertion-that-closes-container-in-DeltaManager-if-sending-an-op-while-concurrently-processing-another-op)
```
Comment (andre4i) [inline BREAKING.md:26]: I don't believe this needs a note in BREAKING.md. As a consumer of the framework, I'm not entirely sure what to do with this information which sounds like something internal to the framework.

File: packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts (hunk 7 above)
Comment (WayneFerrao) [reply packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts:392]: Since the option in DeltaManager is defaulted to false, this error being expected won't be thrown.

File: BREAKING.md (hunk 11 above)
Comment (WayneFerrao) [reply BREAKING.md:905]: I believe this was an unintentional change when I was resolving merge conflicts

File: BREAKING.md (hunk 13 above)
Comment (WayneFerrao) [reply BREAKING.md:26]: I felt that it deserved to stay here as a note that this change will not be optional in the future and the appropriate tests will need to be fixed.

File: packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts [hunk 14]
```diff
@@ -374,6 +383,59 @@ describeNoCompat("SharedMap orderSequentially", (getTestObjectProvider) => {
...
+
+            await provider.ensureSynchronized();
+        };
+
+        itExpects("Should close container when sending an op while processing another op",
+            [{
+                eventName: "fluid:telemetry:Container:ContainerClose",
+                error: "Making changes to data model is disallowed while processing ops.",
+            }], async () => {
+                await setupContainers(testContainerConfig, { "Fluid.Container.ConcurrentOpSend": true });
```
Comment (andre4i) [inline packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts:423]: You should also add a negative test, for when the feature gate is not set.

File: BREAKING.md (hunk 13 above)
Comment (andre4i) [reply BREAKING.md:26]: Can you rephrase this in a way that's relevant to the customer? I'm reading this and seeing some thing about the DeltaManager which I'm not actually using directly and I'm not sure how it would affect my code or my designs.

File: BREAKING.md [hunk 15]
```diff
@@ -36,7 +36,7 @@ This field has been deprecated and will be removed in a future breaking change.
         fileName: string,
 -       createShareLinkType?: ShareLinkTypes,
 +       createShareLinkType?: ShareLinkTypes | ISharingLinkKind,
-    ): 
+    ):
```
Comment (andre4i) [inline BREAKING.md:39]: This file should not have unintentional changes

File: packages/loader/container-loader/src/deltaManager.ts [hunk 16]
```diff
@@ -89,6 +92,15 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
     private pending: ISequencedDocumentMessage[] = [];
     private fetchReason: string | undefined;
 
+    private readonly mc: MonitoringContext;
+
+    // A boolean used to assert that ops are not being sent while processing another op.
+    private opsCurrentlyProcessing: boolean = false;
```
Comment (markfields) [inline packages/loader/container-loader/src/deltaManager.ts:98]: nit: I'd call it `currentlyProcessingOps` I think

File: packages/loader/container-loader/src/deltaManager.ts [hunk 17]
```diff
@@ -295,7 +308,8 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
         };
 
         this.connectionManager = createConnectionManager(props);
-
+        this.mc = loggerToMonitoringContext(logger);
+        this.preventConcurrentOpSend = this.mc.config.getBoolean("Fluid.Container.ConcurrentOpSend") ?? false;
```
Comment (markfields) [inline packages/loader/container-loader/src/deltaManager.ts:312]: nit: I think the output JS is more concise if you do `== true` rather than `?? false`

Comment (markfields) [reply packages/loader/container-loader/src/deltaManager.ts:312]: You can check in typescript playground to quickly compare the JS output.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 18]
```diff
@@ -295,7 +308,8 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
         };
 
         this.connectionManager = createConnectionManager(props);
-
+        this.mc = loggerToMonitoringContext(logger);
```
Comment (markfields) [inline packages/loader/container-loader/src/deltaManager.ts:311]: This is technically a duplicate class variable to `this.logger`, because the MC gets mixed in on the existing logger object.  I think convention is to either ditch `this.logger` (and do `this.mc.logger` throughout), or update `this.logger` to be a property that returns `this.mc.logger`.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 19]
```diff
@@ -764,6 +778,8 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
 
     private processInboundMessage(message: ISequencedDocumentMessage): void {
         const startTime = Date.now();
+        assert(!this.opsCurrentlyProcessing, "Currently processing ops.");
```
Comment (markfields) [inline packages/loader/container-loader/src/deltaManager.ts:781]: nit: I'd say something like "Already" not "Currently", or otherwise explain that the assert means something about reentrancy on this function.  Doesn't really matter because it'll turn into an assert code though.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 20]
```diff
@@ -818,7 +834,7 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
             throw new Error("Attempted to process an inbound message without a handler attached");
         }
         this.handler.process(message);
-
+        this.opsCurrentlyProcessing = false;
```
Comment (markfields) [inline packages/loader/container-loader/src/deltaManager.ts:837]: Logic question here - can you explain why this is reset before emitting the "op" event?  Does that order matter?  Naively I'd expect this to be the absolute last thing in the function, but if there's a reason it should be reset before emitting that event then that would be good to explain in a comment.  Given the nature of this PR, I wouldn't be surprised if there were some subtlety here.

File: packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts [hunk 21]
```diff
@@ -341,10 +341,14 @@ describeNoCompat("SharedMap orderSequentially", (getTestObjectProvider) => {
     beforeEach(() => {
         provider = getTestObjectProvider();
     });
+    afterEach(async () => reset());
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts:344]: can probably just do `afterEach(reset)`?

Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts:344]: And likewise below do `const reset = provider.reset;`

Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts:344]: Also... this may happen automatically somehow...?  Not sure actually.

File: packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts [hunk 22]
```diff
@@ -374,6 +381,74 @@ describeNoCompat("SharedMap orderSequentially", (getTestObjectProvider) => {
...
+        let container2: Container;
+        let dataObject2: ITestFluidObject;
+        let sharedMap2: SharedMap;
+
+        beforeEach(async () => {
+            await reset();
+            provider = getTestObjectProvider();
+        });
+
+        afterEach(async () => reset());
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts:394]: Redundant with the call to `reset` in `beforeEach` (I'd remove that and leave this)

File: packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts [hunk 23]
```diff
@@ -374,6 +381,74 @@ describeNoCompat("SharedMap orderSequentially", (getTestObjectProvider) => {
...
+            }], async () => {
+                await setupContainers(testContainerConfig, { "Fluid.Container.ConcurrentOpSend": true });
+
+                sharedMap.on("valueChanged", (changed, local) => {
+                    if (!local) {
+                        assert.equal(changed.key, "key2", "Incorrect value for key1 in container 1");
+                    }
+                    // Avoid re-entrancy by setting a new key
+                    if (changed.key !== "key2") {
+                        sharedMap2.set("key2", "v2");
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts:429]: This test may be over-complicated?  Not sure, I'm not confident I'm reading it right.  What it looks like to me:

- There are two container instances loaded by the same loader, so using the same DeltaManager.
- This test is simulating the 2nd container firing an op while processing an op from the first container.

Why do you need 2 containers?  And this seems like an unlikely scenario to simulate.  Closest thing I can imagine would be like app code in the main container somehow triggering something that causes that Summarizer to submit an op synchronously.  Maybe it's worth covering that, but maybe not.

I guess now that I write all that out... no harm in testing across that dimension, I was just surprised. So feel free to leave as-is.

**Review by markfields — APPROVED** [review]
Left a bunch of comments, mostly pretty nit-picky - for conversation/instruction more than anything, up to you which ones you take, it looks good to me!  But would be good to get sign-off from Andrei or Vlad too if they've thought more deeply about the problem space.

Also, you probably want to open an item in ADO to track turning on the feature gate and watching the fix get rolled out through our test automation and then some partner deployments.

File: packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts (hunk 21 above)
Comment (WayneFerrao) [reply packages/test/test-end-to-end-tests/src/test/mapEndToEndTests.spec.ts:344]: I run into some reference errors when I try this, namely provider being used before its assigned

File: packages/loader/container-loader/src/deltaManager.ts (hunk 20 above)
Comment (WayneFerrao) [reply packages/loader/container-loader/src/deltaManager.ts:837]: In my understanding, the flag should be reset after the above `process` function is called to signify that op processing was complete.

**Review by vladsud — APPROVED** [review]
Looks good, but I'd add breaking.md and make the check ON by default (using feature gate as a kill bit switch). breaking.md should say that such feature gate will not be available in some future, it's only to give people some time to react
